shared_model/backend/protobuf/block.hpp

```diff
@@ -0,0 +1,135 @@
+      explicit Block(BlockType &&block)
+          : CopyableProto(std::forward<BlockType>(block)),
+            transactions_([this] {
+              std::vector<w<interface::Transaction>> txs;
+              for (const auto &tx : proto_->payload().transactions()) {
+                auto tmp = detail::makePolymorphic<proto::Transaction>(tx);
+                txs.emplace_back(tmp);
+              }
+              return txs;
+            }),
+            blob_([this] { return BlobType(proto_->SerializeAsString()); }),
+            prev_hash_([this] {
+              return HashType(proto_->payload().prev_block_hash());
+            }),
+            signatures_([this] {
+              SignatureSetType sigs;
+              for (const auto &sig : proto_->signatures()) {
+                auto curr = detail::makePolymorphic<proto::Signature>(sig);
+                sigs.insert(curr);
+              }
+              return sigs;
+            }),
+            payload_(detail::makeReferenceGenerator(
+                proto_, &iroha::protocol::Block::payload)),
+            payload_blob_([this] { return makeBlob(*payload_); })
+
+      {}
+
+      Block(const Block &o) : Block(o.proto_) {}
+
+      Block(Block &&o) noexcept : Block(std::move(o.proto_)) {}
+
+      const TransactionsCollectionType &transactions() const override {
+        return *transactions_;
+      }
+
+      interface::types::HeightType height() const override {
+        return payload_->height();
+      }
+
+      const HashType &prevHash() const override {
+        return *prev_hash_;
+      }
+
+      const BlobType &blob() const override {
+        return *blob_;
+      }
+
+      const SignatureSetType &signatures() const override {
+        return *signatures_;
+      }
+
+      bool addSignature(
+          const interface::types::SignatureType &signature) override {
+        if (signatures_->count(signature) > 0) {
+          return false;
+        }
+
+        auto sig = proto_->add_signatures();
+        sig->set_pubkey(crypto::toBinaryString(signature->publicKey()));
+        sig->set_signature(crypto::toBinaryString(signature->signedData()));
+        signatures_.invalidate();
+        return true;
+      }
+
+      interface::types::TimestampType createdTime() const override {
+        return payload_->created_time();
+      }
+
+      TransactionsNumberType txsNumber() const override {
+        return payload_->tx_number();
+      }
+
+      const typename Hashable<Block, iroha::model::Block>::BlobType &payload()
+          const override {
+        return *payload_blob_;
+      }
+
+     private:
+      // lazy
+      template <typename T>
+      using Lazy = detail::LazyInitializer<T>;
+
+      const Lazy<std::vector<w<interface::Transaction>>> transactions_;
+      const Lazy<BlobType> blob_;
+      const Lazy<HashType> prev_hash_;
+      const Lazy<SignatureSetType> signatures_;
+      const Lazy<const iroha::protocol::Block::Payload &> payload_;
```

Comment:
Seems unnecessary to use lazy wrapper for some of the fields

Reply:
I think all these fields are quite "heavy" so lazy should be good for them.